source/server/overload_manager_impl.cc

```diff
@@ -143,8 +142,8 @@ OverloadManagerImpl::OverloadManagerImpl(Event::Dispatcher& dispatcher, Stats::S
   for (const auto& action : config.actions()) {
     const auto& name = action.name();
     ENVOY_LOG(debug, "Adding overload action {}", name);
-    auto result = actions_.emplace(std::piecewise_construct, std::forward_as_tuple(name),
-                                   std::forward_as_tuple(action, stats_scope));
+    // Cannot use in place construction as OverloadAction constructor may throw
+    auto result = actions_.try_emplace(name, OverloadAction(action, stats_scope));
```

Comment:
Are there other places where the emplace and exception interactions come up?  Is there an upstream absl issue tracking making the emplace methods exception safe?  The status quo seems risky.

Reply:
Doing a quick grep, the only place I see the piecewise construction happening with `emplace` is in this file and since we just moved to C++17, we don't seem to have any usage of `try_emplace` yet

Is this absl issue sufficient? https://github.com/abseil/abseil-cpp/issues/388 we can report another

Reply:
Reference the issue in the code.  How did you detect this exception issue?  Should we avoid use of try_emplace until the upstream issue is fixed?

Reply:
adding a note about that issue and the symptoms

we detected this in a unit test failure: https://github.com/envoyproxy/envoy/blob/142b27533ba110afc13651bd75d05b4e25fa1c0d/test/server/overload_manager_impl_test.cc#L359-L382, when the exception is thrown and memory is freed we got an invalid free

With proper test coverage of any constructor that may throw an exception, I would guess this issue would get caught with a test failure, any in-place construction of a type whose constructor does not throw is safe

Reply:
I don't don't know if we have full test coverage for every constructor that can throw an exception which is used as a map value.  I think the main place where this could come up is the control plane doing config validation in constructors while loading config.